### PR TITLE
monitoring: implement owner routing for alerts

### LINKED ISSUE #12010: Approved: Proposal: RFC-189: Support per-team alerts and on-call rotations
### Context

[RFC 189](https://docs.google.com/document/d/1noZf86g2QwTbFNt3XSSbMg36D-_HCptP1thMzZsgt2Q/edit) suggests that we might want engineering teams to own their own alerts. The RFC does not yet thoroughly detail this, but this would likely entail:

* Some way to automatically send alerts to appropriate teams to handle
* Some way to denote ownership of alerts (defined in package `/monitoring`)

#### Simple alert routing

https://github.com/sourcegraph/sourcegraph/pull/11832 adds site-config-based notification definitions via Prometheus Alertmanager. We can extend the `observability.alerts` to accomodate matching on a small set of labels, for example:

```diff
{
  "level": "critical",
  "notifier": {
    "type": "opsgenie",
    "apiKey": "xxx",
    "responders": [ ... ]
  },
+ "onLabels": {
+   "service": [ "git_server", "frontend" ]
+ }
}
```

This alone might be a sufficient (if tedious) way to help teams own their own alerts. It might also cause some alerts to remain unowned. We could also restrict the implied breadth of `onLabels` field from above and just routing fields be a top-level option (example in the next point)

#### Denoting ownership

Ideally, whatever we do to denote ownership should not be Cloud-specific, ie it would be unpleasant to have to generate different alerting for Cloud. An additional required field could be added to our monitoring `Observable`s to give panel a "product area" corresponding to the teams defined in https://github.com/sourcegraph/about/pull/1150, for example:

```diff
{
	Name:            "disk_space_remaining",
	Description:     "disk space remaining by instance",
+	Owner:           OwnerSearch, // "search"
	Query:           `(src_gitserver_disk_space_available / src_gitserver_disk_space_total) * 100`,
	DataMayNotExist: true,
	Critical:        Alert{LessOrEqual: 5},
	PanelOptions:    PanelOptions().LegendFormat("{{instance}}").Unit(Percentage),
	PossibleSolutions: `
		- **Provision more disk space:** Sourcegraph will begin deleting least-used repository clones at 10% disk space remaining which may result in decreased performance, users having to wait for repositories to clone, etc.
	`,
},
```

This would:

* give us useful information on who to contact for an alert when provided alerts by customers (ie via the bug report page)
   * "a lot of alerts in group 'search' are firing, ask someone in the search team"
* make it easier to maintain notifiers by being able to just use site config and the functionality developed to deploy/silence these notifications.
   * enable us to better dogfood alerts (we'll use the same templates, alert timings, etc. as customers) 

Alternative naming for this field: `Team`, `ProductArea`, `Group`

A notifier configuration with the above ownership label them might look like:

```diff
{
  "level": "critical",
  "notifier": {
    "type": "opsgenie",
    "apiKey": "xxx",
    "responders": [ ... ]
  },
+ "owners": [ "search" ]
}
```

### PR DESCRIPTION
**Pull request by bobheadxi**
Adds a new `owners` field to `observability.alerts` as proposed in #12010 for RFC-189. Only alerts that match the specified owners will be received by the associated notifier. Notifiers configured without the `owners` field will continue to receive all alerts of the specified level, even with notifiers that have `owners` present.

image

(outdated screenshot, but close enough)

Closes #12010 - the other half of the proposal was implemented in https://github.com/sourcegraph/sourcegraph/pull/12358

Still thinking about `owners` vs `onOwners` 🤔

### CODE REVIEW

**Comment by pecigonzalo** [comment]
I would use `owners` similarly to how we use `level` and not `onLevel`. Could we add some tests to this to ensure it behaves as expected without doing a full deployment?

**Comment by bobheadxi** [comment]
> I would use owners similarly to how we use level and not onLevel.

🚀 

> Could we add some tests to this to ensure it behaves as expected without doing a full deployment?

Unfortunately the only really relevant thing to test here is how alertmanager actually behaves when sending out alerts, but I'll add something to check on the rendered routes

**Comment by pecigonzalo** [comment]
> I'll add something to check on the rendered routes

Exactly, we should not test Alertmanager itself or that it routes properly, only that we render the expected config given X input.

File: docker-images/prometheus/cmd/prom-wrapper/receivers.go [hunk 1]
```diff
@@ -39,30 +40,56 @@ For possible solutions, please refer to %s`, alertSolutionsURLTemplate)
...
+			owners := strings.Join(alert.Owners, "|")
+			ownerRegexp, err := amconfig.NewRegexp(fmt.Sprintf("^(%s)$", owners))
+			if err != nil {
+				newProblem(fmt.Errorf("failed to apply alert %d: %w", i, err))
+				continue
+			}
+
+			receiverName := fmt.Sprintf("src-%s-on-%s", alert.Level, owners)
+			receiver = &amconfig.Receiver{Name: receiverName}
+			additionalReceviers = append(additionalReceviers, receiver)
```
Comment (slimsag) [inline docker-images/prometheus/cmd/prom-wrapper/receivers.go:77]:
```suggestion
			additionalReceivers = append(additionalReceviers, receiver)
```

File: docker-images/prometheus/cmd/prom-wrapper/receivers.go [hunk 2]
```diff
@@ -39,30 +40,56 @@ For possible solutions, please refer to %s`, alertSolutionsURLTemplate)
...
+			additionalRoutes = append(additionalRoutes, &amconfig.Route{
+				Receiver: receiverName,
+				Match: map[string]string{
+					"level": alert.Level,
+				},
+				MatchRE: amconfig.MatchRegexps{
+					"owner": *ownerRegexp,
+				},
+				Continue: true, // match siblings, so that general-case level receivers still get the alert
```
Comment (slimsag) [inline docker-images/prometheus/cmd/prom-wrapper/receivers.go:86]: What does this comment mean? I'm not sure I follow - could you detail it more?

File: schema/site.schema.json [hunk 3]
```diff
@@ -657,6 +657,13 @@
             "description": "Disable notifications when alerts resolve themselves.",
             "type": "boolean",
             "default": false
+          },
+          "owners": {
+            "description": "Only receive alerts owned by the specific teams - for most use cases, we recommend against using this.",
```
Comment (slimsag) [inline schema/site.schema.json:662]:
```suggestion
            "description": "Do not use. When set, only receive alerts owned by the specified teams. Used for Sourcegraph.com only.",
```

File: docker-images/prometheus/cmd/prom-wrapper/receivers.go (hunk 2 above)
Comment (bobheadxi) [reply docker-images/prometheus/cmd/prom-wrapper/receivers.go]: I clarified it here! 5ed2c890bc75286fa2bc00f98ba2757b69c641ab

**Pull request merged**